src/text-editor.coffee

```diff
@@ -818,114 +818,126 @@ class TextEditor extends Model
+        foldedRows = []
+        rows = [selection.start.row..selection.end.row]
+        if selection.start.row isnt selection.end.row and selection.end.column is 0
+          rows.pop() unless @isFoldedAtBufferRow(selection.end.row)
+
+        # Move line around the fold that is directly above the selection
+        precedingScreenRow = @screenPositionForBufferPosition([selection.start.row]).translate([-1])
+        precedingBufferRow = @bufferPositionForScreenPosition(precedingScreenRow).row
+        if fold = @largestFoldContainingBufferRow(precedingBufferRow)
+          insertDelta = fold.getBufferRange().getRowCount()
         else
-          startRow = row
-          endRow = row
+          insertDelta = 1
 
-        insertPosition = Point.fromObject([startRow - insertDelta])
-        endPosition = Point.min([endRow + 1], @buffer.getEndPosition())
-        lines = @buffer.getTextInRange([[startRow], endPosition])
-        if endPosition.row is lastRow and endPosition.column > 0 and not @buffer.lineEndingForRow(endPosition.row)
-          lines = "#{lines}\n"
+        for row in rows
+          if fold = @displayBuffer.largestFoldStartingAtBufferRow(row)
+            bufferRange = fold.getBufferRange()
+            startRow = bufferRange.start.row
+            endRow = bufferRange.end.row
+            foldedRows.push(startRow - insertDelta)
+          else
+            startRow = row
+            endRow = row
 
-        @buffer.deleteRows(startRow, endRow)
+          insertPosition = Point.fromObject([startRow - insertDelta])
+          endPosition = Point.min([endRow + 1], @buffer.getEndPosition())
+          lines = @buffer.getTextInRange([[startRow], endPosition])
+          if endPosition.row is lastRow and endPosition.column > 0 and not @buffer.lineEndingForRow(endPosition.row)
+            lines = "#{lines}\n"
 
-        # Make sure the inserted text doesn't go into an existing fold
-        if fold = @displayBuffer.largestFoldStartingAtBufferRow(insertPosition.row)
-          @unfoldBufferRow(insertPosition.row)
-          foldedRows.push(insertPosition.row + endRow - startRow + fold.getBufferRange().getRowCount())
+          @buffer.deleteRows(startRow, endRow)
 
-        @buffer.insert(insertPosition, lines)
+          # Make sure the inserted text doesn't go into an existing fold
+          if fold = @displayBuffer.largestFoldStartingAtBufferRow(insertPosition.row)
+            @unfoldBufferRow(insertPosition.row)
+            foldedRows.push(insertPosition.row + endRow - startRow + fold.getBufferRange().getRowCount())
 
-      # Restore folds that existed before the lines were moved
-      for foldedRow in foldedRows when 0 <= foldedRow <= @getLastBufferRow()
-        @foldBufferRow(foldedRow)
+          @buffer.insert(insertPosition, lines)
 
-      @setSelectedBufferRange(selection.translate([-insertDelta]), preserveFolds: true, autoscroll: true)
+        # Restore folds that existed before the lines were moved
```

Comment:
Yeah... my intuition tells me we shouldn't need to mess with unfolding and refolding folds here. Help me understand if I'm wrong.